upgrade/v110to120/upgrade.go

```diff
@@ -260,124 +309,150 @@ func translateVolumeRecurringJobsToLabel(namespace string, lhClient *lhclientset
+	revertLog := run.log
 	defer func() {
-		err = errors.Wrapf(err, upgradeLogPrefix+"cleanup applied volume cron jobs")
-	}()
-
-	log := logrus.WithFields(
-		logrus.Fields{
-			"upgrade":   "cleanup-applied-volume-cron-jobs",
-			"namespace": namespace,
-		},
-	)
-
-	log.Debug(upgradeLogPrefix + "Listing all volumes")
-	volumeList, err := lhClient.LonghornV1beta1().Volumes(namespace).List(context.TODO(), metav1.ListOptions{})
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			log.Debug(upgradeLogPrefix + "Cannot find volumes")
-			return nil
+		if err == nil {
+			run.log.Info(upgradeLogPrefix + "Finished volume cron job cleanup")
 		}
-		return errors.Wrap(err, "failed to list volumes")
-	}
+		err = errors.Wrapf(err, upgradeLogPrefix+"cleanup applied volume cron jobs failed")
+		run.log = revertLog
+	}()
+	run.log = run.log.WithField("action", "cleanup-applied-volume-cron-jobs")
+	run.log.Info(upgradeLogPrefix + "Starting volume cron job cleanup")
 
 	propagation := metav1.DeletePropagationForeground
-	cronJobClient := kubeClient.BatchV1beta1().CronJobs(namespace)
-	for _, v := range volumeList.Items {
-		log.Debugf(upgradeLogPrefix+"Listing all cron jobs for volume %v", v.Name)
-		appliedCronJobROs, err := listVolumeCronJobROs(v.Name, namespace, kubeClient)
+	cronJobClient := run.kubeClient.BatchV1beta1().CronJobs(run.namespace)
+	for _, v := range run.volumes.Items {
+		run.log.Debugf(upgradeLogPrefix+"Listing all cron jobs for volume %v", v.Name)
+		appliedCronJobROs, err := listVolumeCronJobROs(v.Name, run.namespace, run.kubeClient)
 		if err != nil {
 			return errors.Wrapf(err, "failed to list all cron jobs for volume %v", v.Name)
 		}
 		for name := range appliedCronJobROs {
-			log.Infof(upgradeLogPrefix+"Deleting %v cronjob job for %v volume", name, v.Name)
+			run.log.Infof(upgradeLogPrefix+"Deleting %v cronjob job for %v volume", name, v.Name)
 			err := cronJobClient.Delete(context.TODO(), name, metav1.DeleteOptions{PropagationPolicy: &propagation})
 			if err != nil {
 				return errors.Wrapf(err, "failed to delete %v cron job for volume %v", name, v.Name)
 			}
 		}
-		log.Infof(upgradeLogPrefix+"Deleted %v cronjob job for %v volume", len(appliedCronJobROs), v.Name)
+		run.log.Infof(upgradeLogPrefix+"Deleted %v cronjob job for %v volume", len(appliedCronJobROs), v.Name)
+	}
+
+	return nil
+}
+
+func (run *recurringJobUpgrade) createRecurringJobCRs() (err error) {
+	revertLog := run.log
+	defer func() {
+		run.log = revertLog
+		err = errors.Wrapf(err, "create recurringJob CRs failed")
+	}()
+	run.log = run.log.WithField("action", "create-recurringJob-CRs")
+
+	if len(run.recurringJobMapSpec) == 0 {
+		run.log.Debug(upgradeLogPrefix + "Found 0 recurring job")
+		return nil
+	}
+
+	for _, spec := range run.recurringJobMapSpec {
+		run.log.Infof(upgradeLogPrefix+"Creating %v recurring job CR", spec.Name)
+		newRecurringJob := &longhorn.RecurringJob{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: spec.Name,
```

Comment:
I just realized the CR name is stored in the spec as well... Can we remove it in the future or now?

Reply:
I think this PR should stick with the upgrade fix. We can look at the name in spec when we start to do https://github.com/longhorn/longhorn/issues/2865.

Reply:
OK. We can deprecate this spec field in this ticket. But please avoid using it from now on.